ext/drx/drx.c

```diff
@@ -2754,41 +2760,45 @@ drx_try_to_detect_avx512_gather_sequence(void *drcontext, dr_restore_state_info_
                                 advance_state(
                                     &detect_state,
                                     DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
-                                    &allow_for_unknown_instr_count);
+                                    &skip_unknown_instr_count);
                             }
                         }
                     }
                 }
             }
-            if (allow_for_unknown_instr_inc(&allow_for_unknown_instr_count))
-                break;
-            advance_state(&detect_state, DRX_DETECT_RESTORE_AVX512_GATHER_EVENT_STATE_0,
-                          &allow_for_unknown_instr_count);
+            skip_unknown_instr_inc(&detect_state, &skip_unknown_instr_count);
             break;
         default: ASSERT(false, "internal error: invalid state.");
         }
     }
+    return true;
 }
 
 static bool
 drx_event_restore_state(void *drcontext, bool restore_memory,
                         dr_restore_state_info_t *info)
 {
     instr_t inst;
+    bool success = true;
     if (info->fragment_info.cache_start_pc == NULL)
         return true; /* fault not in cache */
+    if (!expand_scatter_gather_drreg_initialized) {
+        /* Nothing to do if nobody had never called expand_scatter_gather() before. */
+        return true;
+    }
     instr_init(drcontext, &inst);
     byte *pc = decode(drcontext, dr_fragment_app_pc(info->fragment_info.tag), &inst);
     if (pc != NULL) {
         if (instr_is_gather(&inst)) {
-            /* XXX i#2985: if !info->fragment_info.app_code_consistent, we run the state
-             * machine anyway.
-             */
+            if (!info->fragment_info.app_code_consistent) {
+                /* Can't verify application code. Let the translation fail. */
```

Comment:
Per your observation, this does not seem safe: this will fail selfmod or deleted fragments having nothing to do w/ scatter-gather and could cause fatal problems where before everything worked fine.

Reply:
I've added XXX comment and return true. I wonder if it's better to just run the state machine anyway.